
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Inputting User Data using the Hubble Advanced Spectral Products Script &#8212; HST Notebooks</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.8ecb98da25f57f5357bf6f572d296f466b2cfe2517ffebfabe82451661e28f02.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
    <link rel="stylesheet" type="text/css" href="../../../_static/table.css?v=cd6c26fb" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/HASP/CoaddTutorial/CoaddTutorial';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Scaling Flux while using the Hubble Advanced Spectral Products Script" href="../FluxScaleTutorial/FluxScaleTutorial.html" />
    <link rel="prev" title="Installing the Hubble Advanced Spectral Products Script" href="../Setup/Setup.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/stsci_logo2.png" class="logo__image only-light" alt="HST Notebooks - Home"/>
    <script>document.write(`<img src="../../../_static/stsci_logo2.png" class="logo__image only-dark" alt="HST Notebooks - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../index.html">
                    STScI HST Notebook Repository HQ
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">ACS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../ACS/README.html">ACS Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_findsat_mrt/acs_findsat_mrt_example.html">Satellite trail detection in ACS/WFC data using acstools.findsat_mrt</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_saturation_trails/acs_saturation_trails.html">ACS Linearity with Saturated Stars</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_cte_forward_model/acs_cte_forward_model_example.html">Pixel-based ACS/WFC CTE Forward Model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_subarrays/acs_subarrays.html">ACS Image Reduction for Subarray Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_sbc_dark_analysis/acs_sbc_dark_analysis.html">SBC Dark Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_reduction/acs_reduction.html">ACS/WFC Image Reduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_pixel_area_maps/acs_pixel_area_maps.html">ACS Pixel Area Maps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_polarization_tools/acs_polarization_tools.html">Using ACS Polarization Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_focus_diverse_epsfs/acs_focus_diverse_epsfs.html">Focus Diverse ePSFs for ACS/WFC</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../ACS/acs_exception_report/exception_report_checks.html">HST Exception Report - Investigate your ACS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../ACS/hst_orbits_ephem/hst_orbits_ephem.html">Getting HST Ephemeris and Related Data for an Observation</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">COS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../COS/README.html">COS Notebooks - testing update 2!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../COS/Setup/Setup.html">Setting up your computer environment for working with COS data</a></li>





<li class="toctree-l1"><a class="reference internal" href="../../COS/DataDl/DataDl.html">Downloading COS Data</a></li>




<li class="toctree-l1"><a class="reference internal" href="../../COS/ViewData/ViewData.html">Viewing COS Data</a></li>





<li class="toctree-l1"><a class="reference internal" href="../../COS/AsnFile/AsnFile.html">Modifying or Creating an Association File</a></li>





<li class="toctree-l1"><a class="reference internal" href="../../COS/CalCOS/CalCOS.html">Running the COS Data Pipeline (<code class="docutils literal notranslate"><span class="pre">CalCOS</span></code>)</a></li>






<li class="toctree-l1"><a class="reference internal" href="../../COS/SplitTag/SplitTag.html">Splitting COS Exposures into sub-exposures with <code class="docutils literal notranslate"><span class="pre">splittag</span></code></a></li>







<li class="toctree-l1"><a class="reference internal" href="../../COS/DayNight/DayNight.html">Filtering out COS Data taken during the Day or Night</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../COS/LSF/LSF.html">Working with the COS Line Spread Function (LSF)</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../COS/Extract/Extract.html">Editing the extraction boxes in a spectral extraction file (<code class="docutils literal notranslate"><span class="pre">XTRACTAB</span></code>)</a></li>







<li class="toctree-l1"><a class="reference internal" href="../../COS/ExceptionReport/ExceptionReport.html">What to do when you get an Exception Report for COS</a></li>







</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">DrizzlePac</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/README.html">DrizzlePac Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/using_updated_astrometry_solutions/using_updated_astrometry_solutions.html">Improving Astrometry Using Alternate WCS Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_to_catalogs/align_to_catalogs.html">Aligning HST images to an Absolute Reference Catalog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_sparse_fields/align_sparse_fields.html">Aligning Deep Exposures of Sparse Fields</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_multiple_visits/align_multiple_visits.html">Optimizing Image Alignment for Multiple HST Visits</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/use_ds9_regions_in_tweakreg/use_ds9_regions_in_tweakreg.html">TWEAKREG Tips - Using DS9 Regions for Source Inclusion/Exclusion</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/optimize_image_sampling/optimize_image_sampling.html">Optimizing the Image Sampling</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/align_mosaics/align_mosaics.html">Aligning HST Mosaics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/sky_matching/sky_matching.html">Sky Matching for HST Mosaics <a id="top"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/mask_satellite/mask_satellite.html">Satellite Trail Masking Techniques</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../DrizzlePac/drizzle_wfpc2/drizzle_wfpc2.html">Drizzling new WFPC2 FLT data products with chip-normalization   </a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">HASP</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../Setup/Setup.html">Installing the Hubble Advanced Spectral Products Script</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Inputting User Data using the Hubble Advanced Spectral Products Script</a></li>

<li class="toctree-l1"><a class="reference internal" href="../FluxScaleTutorial/FluxScaleTutorial.html">Scaling Flux while using the Hubble Advanced Spectral Products Script</a></li>

<li class="toctree-l1"><a class="reference internal" href="../DataDiagnostic/DataDiagnostics.html">HASP Data Diagnostic Notebook</a></li>




<li class="toctree-l1"><a class="reference internal" href="../WavelengthAdjustment/WavelengthAdjustment.html">Wavelength Adjustments for Running the Hubble Advanced Spectral Products (HASP) Script</a></li>

<li class="toctree-l1"><a class="reference internal" href="../COSLifetimePositions/COSLifetimePositions.html">Combining COS Data from Multiple Lifetime Positions and Central Wavelengths</a></li>

</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">NICMOS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../NICMOS/nicmos_unit_conversion/nicmos_unit_conversion.html">Programmatic Replacements for NICMOS Units Conversion Form</a></li>


</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">STIS</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../STIS/README.html">STIS Notebooks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/cross-correlation/cross-correlation.html">Correcting for Missing Wavecals with Cross-Correlation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/custom_ccd_darks/custom_ccd_darks.html">Custom CCD Darks <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/drizpac_notebook/STIS_DrizzlePac_Tutorial.html">STIS DrizzlePac Tutorial <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/CoronagraphyViz/STIS_Coronagraphy_Visualization_v2.html">STIS Coronagraphy Visualization Tool (v2)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/calstis/calstis_2d_ccd.html">Calstis 2-D CCD Data Reduction <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/target_acquisition/target_acquisition.html">Evaluating STIS Target Acquisitions <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/view_data/view_data.html">Viewing STIS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/extraction/1D_Extraction.html">1D Spectra Extraction <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/low_count_uncertainties/Low_Count_Uncertainties.html">Low Count Uncertainties in STIS <a class="tocSkip"></a></a></li>
<li class="toctree-l1"><a class="reference internal" href="../../STIS/contrast_sensitivity/STIS_Coronagraphic_Observation_Feasibility.html">STIS Coronagraphic Observation Feasibility</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">WFC3</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../WFC3/README.html">WFC3 Notebooks</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/general_tools.html">General Tools</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/image_displayer_analyzer/wfc3_image_displayer_analyzer.html">WFC3 Image Displayer &amp; Analyzer</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/exception_report/wfc3_exception_report.html">Exception Report Checklist - WFC3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/calwf3_v1.0_cte/calwf3_with_v1.0_PCTE.html">Processing WFC3/UVIS Data with <code class="docutils literal notranslate"><span class="pre">calwf3</span></code> Using the v1.0 CTE-Correction</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/persistence/wfc3_ir_persistence.html">Masking Persistence in WFC3/IR Images</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/uvis_g280_transit/G280_Exoplanet_Transits.html">Analyzing WFC3/UVIS G280 Exoplanet Transit Observations</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/ir_tvb.html">WFC3/IR Time Variable Background (TVB)</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/ir_ima_visualization/IR_IMA_Visualization_with_an_Example_of_Time_Variable_Background.html">WFC3/IR IMA Visualization Tools with An Example of Time Variable Background</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/calwf3_recalibration/calwf3_recal_tvb.html">Manual Recalibration with calwf3: Turning off the IR Linear Ramp Fit</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/tvb_flattenramp/TVB_flattenramp_notebook.html">Correcting for Helium Line Emission Background in IR Exposures using the “Flatten-Ramp” Technique</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/ir_scattered_light_calwf3_corrections/Correcting_for_Scattered_Light_in_IR_Exposures_Using_calwf3_to_Mask_Bad_Reads.html">Correcting for Scattered Light in WFC3/IR Exposures: Using <code class="docutils literal notranslate"><span class="pre">calwf3</span></code> to Mask Bad Reads</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/ir_scattered_light_manual_corrections/Correcting_for_Scattered_Light_in_IR_Exposures_by_Manually_Subtracting_Bad_Reads.html">Correcting for Scattered Light in WFC3/IR Exposures: Manually Subtracting Bad Reads</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/photometry.html">Photometry</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/filter_transformations/filter_transformations.html">WFC3/UVIS Filter Transformations with stsynphot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/flux_conversion_tool/flux_conversion_tool.html">Flux Unit Conversions with synphot and stsynphot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/photometry_examples/phot_examples.html">Synthetic Photometry Examples for WFC3</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/uvis_time_dependent_photometry/uvis_timedep_phot.html">WFC3/UVIS Time-dependent Photometry</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/zeropoints/zeropoints.html">Calculating WFC3 Zeropoints with STSynphot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/uvis_pam_corrections/WFC3_UVIS_Pixel_Area_Map_Corrections_for_Subarrays.html">WFC3/UVIS Pixel Area Map Corrections for Subarrays</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../WFC3/point_spread_function.html">Point Spread Function (PSF)</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/point_spread_function/hst_point_spread_function.html">HST WFC3 Point Spread Function Modeling   </a></li>
<li class="toctree-l2"><a class="reference internal" href="../../WFC3/mast_api_psf/download_psf_cutouts.html">Downloading WFC3 and WFPC2 PSF Cutouts from MAST</a></li>
</ul>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/spacetelescope/hst_notebooks/blob/main/notebooks/HASP/CoaddTutorial/CoaddTutorial.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch on Colab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img alt="Colab logo" src="../../../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="initThebeSBT()"
  class="btn btn-sm btn-launch-thebe dropdown-item"
  title="Launch Thebe"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-play"></i>
  </span>
<span class="btn__text-container">Live Code</span>
</button>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/spacetelescope/hst_notebooks" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/spacetelescope/hst_notebooks/issues/new?title=Issue%20on%20page%20%2Fnotebooks/HASP/CoaddTutorial/CoaddTutorial.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/notebooks/HASP/CoaddTutorial/CoaddTutorial.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Inputting User Data using the Hubble Advanced Spectral Products Script</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Inputting User Data using the Hubble Advanced Spectral Products Script</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#this-notebook-is-designed-to-walk-you-through-downloading-and-using-the-hubble-advanced-spectral-products-hasp-co-add-script"><span style="font-weight:normal">This Notebook is designed to walk you through downloading and using the <strong>Hubble Advanced Spectral Products (HASP)</strong> co-add script.</span></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-goals">Learning Goals:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">0. Introduction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#imports">Imports</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-hst-spectroscopic-data">1. Downloading HST Spectroscopic Data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#using-astroquery-to-download-stis-data">1.1 Using <code class="docutils literal notranslate"><span class="pre">Astroquery</span></code> to Download STIS Data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-the-co-add-script">2. Running the Co-add Script</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#working-with-co-added-data-products">3. Working with Co-added Data Products</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#understanding-the-output-files">3.1 Understanding the Output Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#viewing-the-co-added-data">3.2 Viewing the Co-added Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#putting-it-all-together-with-cos-data">3.3 Putting it all together with COS data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-hasp-data-products">4. Downloading HASP Data Products</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-data-products-using-astroquery">4.1 Downloading Data Products using <code class="docutils literal notranslate"><span class="pre">astroquery</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#changing-the-threshold-flag">5. Changing the Threshold Flag</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-the-data">5.1 Downloading the Data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-the-co-add-script-with-different-threshold-values">5.2 Running the Co-add Script with Different Threshold Values</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analyzing-the-different-co-added-spectra-of-different-threshold-values">5.3 Analyzing the Different Co-added Spectra of Different Threshold Values</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#looking-at-snr">Looking at SNR:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#looking-at-flux">Looking at Flux:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#congrats-on-completing-the-notebook">Congrats on completing the notebook!</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#there-are-more-tutorial-notebooks-for-custom-co-addition-cases-in-this-repo-check-them-out">There are more tutorial notebooks for custom co-addition cases in this repo, check them out!</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#about-this-notebook">About this Notebook</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citations">Citations</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div style='background-color: #fff3cd; border: 1px solid #ffeaa7; border-radius: 4px; padding: 15px; margin: 10px 0; border-left: 4px solid #f39c12;'>
<h3 style='color: #856404; margin-top: 0;'>⚠️ EXECUTION WARNING</h3>
<p style='color: #856404; margin-bottom: 0;'><strong>This notebook may not execute properly in the current environment.</strong></p>
<p style='color: #856404; margin-bottom: 0;'>Some cells may have failed during automated testing. Please review the notebook content and test manually before use.</p>
<p style='color: #856404; margin-bottom: 0; font-size: 0.9em;'><em>Generated during CI/CD pipeline - some outputs may be incomplete or missing.</em></p>
</div>
<section class="tex2jax_ignore mathjax_ignore" id="inputting-user-data-using-the-hubble-advanced-spectral-products-script">
<h1>Inputting User Data using the Hubble Advanced Spectral Products Script<a class="headerlink" href="#inputting-user-data-using-the-hubble-advanced-spectral-products-script" title="Link to this heading">#</a></h1>
<section id="this-notebook-is-designed-to-walk-you-through-downloading-and-using-the-hubble-advanced-spectral-products-hasp-co-add-script">
<h2><span style="font-weight:normal">This Notebook is designed to walk you through downloading and using the <strong><a class="reference external" href="https://archive.stsci.edu/missions-and-data/hst/hasp">Hubble Advanced Spectral Products (HASP)</a></strong> co-add script.</span><a class="headerlink" href="#this-notebook-is-designed-to-walk-you-through-downloading-and-using-the-hubble-advanced-spectral-products-hasp-co-add-script" title="Link to this heading">#</a></h2>
</section>
<section id="learning-goals">
<h2>Learning Goals:<a class="headerlink" href="#learning-goals" title="Link to this heading">#</a></h2>
<p>By the end of this tutorial, you will learn how to:</p>
<ul class="simple">
<li><p>Setup a <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</p></li>
<li><p>Download the HASP wrapper script</p></li>
<li><p>Use <code class="docutils literal notranslate"><span class="pre">astroquery.mast</span></code> to download COS and STIS data</p></li>
<li><p>Run the co-add script</p></li>
<li><p>Examine the co-added output</p></li>
<li><p>Run the scipt with multiple threshold values and examine output</p></li>
</ul>
</section>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Link to this heading">#</a></h2>
<p><strong>0. <a class="reference internal" href="#introduction"><span class="xref myst">Introduction</span></a></strong></p>
<p><strong>1. <a class="reference internal" href="#datadownload"><span class="xref myst">Downloading HST Spectroscopic Data</span></a></strong></p>
<p>- 1.1 <a class="reference internal" href="#stisdownload"><span class="xref myst">Using <code class="docutils literal notranslate"><span class="pre">astroquery</span></code> to Download STIS Data</span></a></p>
<p><strong>2. <a class="reference internal" href="#runscript"><span class="xref myst">Running the Co-add Script</span></a></strong></p>
<p><strong>3. <a class="reference internal" href="#workwithoutput"><span class="xref myst">Working with Co-added Data Products</span></a></strong></p>
<p>- 3.1 <a class="reference internal" href="#inspectoutput"><span class="xref myst">Inspecting the Output Files</span></a></p>
<p>- 3.2 <a class="reference internal" href="#viewstis"><span class="xref myst">Viewing Co-added STIS Data</span></a></p>
<p>- 3.3 <a class="reference internal" href="#cos"><span class="xref myst">Putting it all together with COS Data</span></a></p>
<p><strong>4. <a class="reference internal" href="#download_prods"><span class="xref myst">Downloading HASP Data Products</span></a></strong></p>
<p>- 4.1 <a class="reference internal" href="#hasp_query"><span class="xref myst">Downloading Data Products using <code class="docutils literal notranslate"><span class="pre">astroquery</span></code></span></a></p>
<p><strong>5. <a class="reference internal" href="#threshold"><span class="xref myst">Changing the Threshold Flag</span></a></strong></p>
<p>- 5.1 <a class="reference internal" href="#downloadthresh"><span class="xref myst">Using <code class="docutils literal notranslate"><span class="pre">Astroquery</span></code> to Download Additional Data</span></a></p>
<p>- 5.2 <a class="reference internal" href="#threshscriptrun"><span class="xref myst">Running the Co-add Script with Multiple Threshold Values</span></a></p>
<p>- 5.3 <a class="reference internal" href="#analyzethresh"><span class="xref myst">Analyzing the Co-added Spectra of Different Threshold Values</span></a></p>
<p><a id = introduction></a></p>
</section>
<section id="introduction">
<h2>0. Introduction<a class="headerlink" href="#introduction" title="Link to this heading">#</a></h2>
<p>The <a class="reference external" href="https://github.com/spacetelescope/hasp">Hubble Advanced Spectral Products (HASP) code</a> is a script that co-adds spectra of the same target within programs. This software is able to co-add data taken with the spectrographs onboard the <a class="reference external" href="https://www.stsci.edu/hst">Hubble Space Telescope (HST)</a>; the <a class="reference external" href="https://www.stsci.edu/hst/instrumentation/stis">Space Telescope Imaging Spectrograph (STIS)</a> and the <a class="reference external" href="https://www.stsci.edu/hst/instrumentation/cos">Cosmic Origins Spectrograph (COS)</a>. The <a class="reference external" href="https://archive.stsci.edu/missions-and-data/hst/hasp">Hubble Spectroscopic Legacy Archive (HSLA)</a> uses this script to co-add these instruments’ data from <a class="reference external" href="https://archive.stsci.edu/">The Mikulski Archive for Space Telescopes (MAST)</a> to create high-quality spectra with a broad wavelength coverage (whenever possible from the ultraviolet to the near-infrared) that is publicly available for the scientific community. These custom co-addition notebooks will instruct users on how to produce their own co-adds in cases where the MAST archive data needs special processing or is rejected by the default filters used in the co-add script.</p>
<p>The script first co-adds the observations for each grating for a given program, then it combines all gratings for the observation set. Finally, it co-adds the spectra of each observation set in the program to produce a fully co-added spectra for each target in a program. Please check out the <a class="reference external" href="https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/cos/documentation/instrument-science-reports-isrs/_documents/ISR2024-01.pdf">COS 2024-01 ISR</a> for more information about HASP.</p>
<p>This notebook will show users how to download data from MAST, run the co-add script, understand the output files and inspect the abutted data by plotting flux as a function of wavelength. It will also show users how to change the flux threshold flag.</p>
<p><strong>Please check out our <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/HASP/Setup">Setup.ipynb</a> notebook before running this tutorial to learn how to install and run the co-add code.</strong></p>
<section id="imports">
<h3>Imports<a class="headerlink" href="#imports" title="Link to this heading">#</a></h3>
<p>We will be using multiple libraries to retrieve and analyze data. We will use:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Path.pathlib</span></code> to create product and data directories</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">astroquery.mast</span> <span class="pre">Observations</span></code> to download COS and STIS data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">shutil</span></code> to perform directory and file operations</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">os</span></code> to interact with the operating system</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">astropy.io</span> <span class="pre">fits</span></code> to work with FITS files</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">matplotlib.pyplot</span></code> to plot abutted spectra</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">glob</span></code> to work with multiple files in our directories</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">subprocesses</span></code> to run our script in the notebook with varying threshold flag values</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">numpy</span></code> to help analyze our data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">scipy.interpolate</span> <span class="pre">interp1d</span></code> to interpolate our data</p></li>
</ul>
<p>We recommend creating a HASP-specific <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment when co-adding spectra. You can checkout our <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/HASP/Setup">Setup.ipynb</a> notebook to create such an environment. Alternatively, you can also download the required dependencies to run this notebook with the terminal command:</p>
<p><code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">-r</span> <span class="pre">requirements.txt</span></code></p>
<p>This will download the dependencies that are necessary to run this current notebook. Let’s import all of our packages that we will use in this notebook and print our <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment by running the next cell:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astroquery.mast</span><span class="w"> </span><span class="kn">import</span> <span class="n">Observations</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">glob</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">glob</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">astropy.io</span><span class="w"> </span><span class="kn">import</span> <span class="n">fits</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">subprocess</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.interpolate</span><span class="w"> </span><span class="kn">import</span> <span class="n">interp1d</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Currently active conda environment:&quot;</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;CONDA_PREFIX&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<p>To do our tutorial, we will create data folders that will contain downloaded data from MAST (one for the STIS and the another for COS). We will also create products folders to contain the HASP script output, a.k.a the co-added spectra. We will have one folder for STIS and another for COS too.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Creating the data download directories for COS and STIS</span>
<span class="n">stis_data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./stis_data/&quot;</span><span class="p">)</span>
<span class="n">cos_data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./cos_data/&quot;</span><span class="p">)</span>

<span class="c1"># Creating the products directory to hold the output</span>
<span class="n">stis_products_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./stis_products/&quot;</span><span class="p">)</span>
<span class="n">cos_products_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./cos_products/&quot;</span><span class="p">)</span>

<span class="c1"># If the directory doesn&#39;t exist, then create it</span>
<span class="n">stis_data_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cos_data_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">stis_products_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">cos_products_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><a id = datadownload></a></p>
</section>
</section>
<section id="downloading-hst-spectroscopic-data">
<h2>1. Downloading HST Spectroscopic Data<a class="headerlink" href="#downloading-hst-spectroscopic-data" title="Link to this heading">#</a></h2>
<p>Now that we have a <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment created and the co-add code downloaded, we can start downloading data using <code class="docutils literal notranslate"><span class="pre">Observations</span></code> class from the Python package <code class="docutils literal notranslate"><span class="pre">astroquery.mast</span></code>. Here we will download two datasets, one taken with STIS and the other taken with COS.</p>
<p><a id = stisdownload></a></p>
<section id="using-astroquery-to-download-stis-data">
<h3>1.1 Using <code class="docutils literal notranslate"><span class="pre">Astroquery</span></code> to Download STIS Data<a class="headerlink" href="#using-astroquery-to-download-stis-data" title="Link to this heading">#</a></h3>
<p>We will be downloading STIS data for the white dwarf <a class="reference external" href="http://simbad.u-strasbg.fr/simbad/sim-id?Ident=GD+++71">GD71</a>; this object is a well-known primary white dwarf standard. We will specifically download data from Program 7656, which has observations of GD71 using the gratings <code class="docutils literal notranslate"><span class="pre">G230L</span></code> and <code class="docutils literal notranslate"><span class="pre">G140L</span></code>.</p>
<p>We can start with querying the MAST database for the STIS program’s data. This will give us a list of <em>all</em> observations for the program.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Querying the data on MAST for our target</span>
<span class="n">gd71_query</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
    <span class="n">proposal_id</span><span class="o">=</span><span class="mi">7656</span><span class="p">,</span>
    <span class="n">target_name</span><span class="o">=</span><span class="s2">&quot;GD71&quot;</span><span class="p">,</span>
    <span class="n">dataproduct_type</span><span class="o">=</span><span class="s2">&quot;SPECTRUM&quot;</span><span class="p">,</span>
    <span class="n">provenance_name</span><span class="o">=</span><span class="s2">&quot;CALSTIS&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now that we have queried the observations for Program 7656, we can get a list that contains the data products for these observations:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Getting a product list for our query</span>
<span class="n">gd71_products</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
    <span class="n">gd71_query</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s print out this list to see the associated data files:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">gd71_products</span><span class="p">[</span><span class="s2">&quot;productFilename&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<p>As you can see, we have a very long list of different type of data products for our program. Luckily, we don’t need all of these files to run the wrapper. We only need to download the following COS and/or STIS files:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://hst-docs.stsci.edu/cosdhb/chapter-2-cos-data-files/2-4-cos-data-products#:~:text=in%20the%20association.-,One%2DDimensional%20Extracted%20Spectra%20(x1d%2C%20x1dsum),-The%20COS%20pipeline">X1D</a> - the one-dimensional extracted product spectra.</p></li>
<li><p><a class="reference external" href="https://hst-docs.stsci.edu/stisdhb/chapter-2-stis-data-structure/2-2-types-of-stis-files#:~:text=corrected%20imaging%20data.-,_sx1,-table">SX1</a> - the one-dimensional extracted spectra from combined or cosmic-ray rejected images. This file is only produced with STIS data.</p></li>
</ul>
<p>We will specify that we want to download <em>only</em> these files with the <code class="docutils literal notranslate"><span class="pre">productSubGroupDescription</span></code> parameter. We will also specify the directory that will contain the downloaded data products. Below, we download the STIS files for the progam.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
    <span class="n">gd71_products</span><span class="p">,</span>
    <span class="n">download_dir</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">stis_data_dir</span><span class="p">),</span>
    <span class="n">productSubGroupDescription</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;X1D&quot;</span><span class="p">,</span> <span class="s2">&quot;SX1&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>When we downloaded the data using <code class="docutils literal notranslate"><span class="pre">astroquery</span></code>, it created a directory <code class="docutils literal notranslate"><span class="pre">./stis_data/mastDownload/HST</span></code>, with separate folders for each different dataset ID. The script will need all of our newly downloaded data product files in a single directory, so we must move all STIS files to our <code class="docutils literal notranslate"><span class="pre">./stis_data</span></code> directory. We will create a function to consolidate our data, since we will be utilizing it for two additional datasets later on in the tutorial.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">consolidate_files</span><span class="p">(</span><span class="n">data_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Consolidate all files to single directory; necessary for HASP script run.</span>
<span class="sd">    ---------------</span>
<span class="sd">    Input:</span>
<span class="sd">    str data_path : ./mastDownload/HST folders paths; files to be moved here</span>
<span class="sd">    ---------------</span>
<span class="sd">    Output:</span>
<span class="sd">    None. Files moved to data_path. ./mastDownload/HST directory is deleted.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># The path to all obs_id folders</span>
    <span class="n">mast_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s2">/mastDownload/HST/&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="c1"># Check if mastDownload exists</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">mast_path</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Directory </span><span class="si">{</span><span class="n">mast_path</span><span class="si">}</span><span class="s2"> doesn&#39;t exist.&quot;</span><span class="p">)</span>
            <span class="k">return</span>

        <span class="c1"># Get a list of the obs_id paths in mastDownload</span>
        <span class="n">obs_id_dirs</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">mast_path</span><span class="p">)</span>

        <span class="c1"># Iterate through each obs_id folder and move the files</span>
        <span class="k">for</span> <span class="n">obs_id</span> <span class="ow">in</span> <span class="n">obs_id_dirs</span><span class="p">:</span>
            <span class="n">obs_id_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mast_path</span><span class="p">,</span> <span class="n">obs_id</span><span class="p">)</span>
            <span class="n">files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">obs_id_path</span> <span class="o">+</span> <span class="s2">&quot;/*fits&quot;</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
                <span class="n">file_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
                <span class="n">new_path</span> <span class="o">=</span> <span class="n">data_path</span> <span class="o">/</span> <span class="n">file_path</span><span class="o">.</span><span class="n">name</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">new_path</span><span class="p">)</span>

        <span class="c1"># Now we can remove the mastDownload directory</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">mast_path</span><span class="p">):</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s2">/mastDownload&quot;</span><span class="p">)</span>

    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;An error occurred: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now, using the function to move our STIS files to a single directory:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">consolidate_files</span><span class="p">(</span><span class="n">stis_data_dir</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we can run the co-add script!</p>
<p><a id = runscript></a></p>
</section>
</section>
<section id="running-the-co-add-script">
<h2>2. Running the Co-add Script<a class="headerlink" href="#running-the-co-add-script" title="Link to this heading">#</a></h2>
<p>Now that we’ve downloaded the GD71 STIS data, we can run the co-add script. Currently, the co-add code abuts spectra for a single program. Run the script by using the next cell’s command.</p>
<p><strong>Note: Make sure that you are in the <code class="docutils literal notranslate"><span class="pre">hasp-env</span></code> <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment that we created at the beginning of the <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/HASP/Setup">Setup.ipynb</a> notebook.</strong></p>
<p>The <code class="docutils literal notranslate"><span class="pre">-i</span></code> parameter is the input directory (i.e, where the FITS files are located). <code class="docutils literal notranslate"><span class="pre">-o</span></code> is the directory that will contain the newly created co-added products. Note that if you want to exclude certain data files from the co-add, you can just remove them from the input directory. There is more information about this (and the other flags) in our <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/HASP/Setup">Setup.ipynb</a> notebook.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>swrapper<span class="w"> </span>-i<span class="w"> </span>./stis_data<span class="w"> </span>-o<span class="w"> </span>./stis_products
</pre></div>
</div>
</div>
</div>
<p>We have now created the co-added products for Program 7656 using the wrapper!</p>
<p><a id = workwithoutput></a></p>
</section>
<section id="working-with-co-added-data-products">
<h2>3. Working with Co-added Data Products<a class="headerlink" href="#working-with-co-added-data-products" title="Link to this heading">#</a></h2>
<p>With the newly co-added files in the <code class="docutils literal notranslate"><span class="pre">./stis_products/output</span></code> directory, we can begin to inspect the data.</p>
<p><a id = inspectoutput></a></p>
<section id="understanding-the-output-files">
<h3>3.1 Understanding the Output Files<a class="headerlink" href="#understanding-the-output-files" title="Link to this heading">#</a></h3>
<p><strong><em>The following information about the output file naming conventions is in our <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/HASP/Setup">Setup.ipynb</a> notebook. If you’re already familiar with this, you can skip to <a class="reference internal" href="#viewstis"><span class="xref myst">Section 4.2</span></a>, where we view our co-added spectra.</em></strong></p>
<p>Let’s look at the <code class="docutils literal notranslate"><span class="pre">./stis_products/output</span></code> directory to look at the newly abutted spectra. Currently, the script outputs abutted products for a single program.</p>
<p>The script produces multiple different files with abutted spectra. Currently, the script outputs abutted products for a single program. It first creates co-added spectra for each grating of a single observation set:</p>
<p><code class="docutils literal notranslate"><span class="pre">hst_programID_instrument_targetname_grating_obset_cspec.fits</span></code></p>
<p>It then co-adds the spectra of all gratings for a single observation set:</p>
<p><code class="docutils literal notranslate"><span class="pre">hst_programID_instrument_targetname_allGratings_obset_cspec.fits</span></code></p>
<p>Finally, it co-adds all abutted observation sets’ spectra to create a final co-added product for a single target:</p>
<p><code class="docutils literal notranslate"><span class="pre">hst_programID_instrument_targetname_allGratings_cspec.fits</span></code></p>
<p>An example of this is below. These filenames are the output files for our STIS GD71 dataset that is co-added in this notebook. Here, the <code class="docutils literal notranslate"><span class="pre">programID</span></code> is <code class="docutils literal notranslate"><span class="pre">7656</span></code>, the <code class="docutils literal notranslate"><span class="pre">instrument</span></code> is <code class="docutils literal notranslate"><span class="pre">STIS</span></code>, and the <code class="docutils literal notranslate"><span class="pre">targetname</span></code> is <code class="docutils literal notranslate"><span class="pre">gd71</span></code>.</p>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Filename</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">hst_7656_stis_gd71_g140l_o4a520_cspec.fits</span></code></p></td>
<td><p>Co-adding all <code class="docutils literal notranslate"><span class="pre">G140L</span></code> observations for the observation set, <code class="docutils literal notranslate"><span class="pre">O4A520</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">hst_7656_stis_gd71_g140l-g230l-g430l-g750l_o4a520_cspec.fits</span></code></p></td>
<td><p>Co-adding all observations taken at every grating for the observation set, <code class="docutils literal notranslate"><span class="pre">O4A520</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">hst_7656_stis_gd71_g140l-g230l-g430l-g750l_o4a5_cspec.fits</span></code></p></td>
<td><p>Co-adding all GD71 observations at each grating for this program, <code class="docutils literal notranslate"><span class="pre">O4A5</span></code>.</p></td>
</tr>
</tbody>
</table>
</div>
<p><em><strong>Note: HST file naming conventions use a combination of three letters and/or numbers to have a unique association between a PI’s proposal ID and program ID, meaning that <code class="docutils literal notranslate"><span class="pre">o4a5</span></code> at the end of <code class="docutils literal notranslate"><span class="pre">hst_7656_stis_gd71_g140l-g230l-g430l-g750l_o4a5_cspec.fits</span></code> is essentially the program ID for our example. Check out more information on the <a class="reference external" href="https://archive.stsci.edu/hlsp/ipppssoot.html">MAST HST file naming convention page</a></strong></em></p>
<p><a id = viewstis></a></p>
</section>
<section id="viewing-the-co-added-data">
<h3>3.2 Viewing the Co-added Data<a class="headerlink" href="#viewing-the-co-added-data" title="Link to this heading">#</a></h3>
<p>Let’s take a look at the co-added spectra that we just created. We will create a plot of flux as a function of wavelength using <code class="docutils literal notranslate"><span class="pre">matplotlib.pyplot</span></code>.</p>
<p>With the current version of the HASP script, the fully abutted filename should be:</p>
<p><code class="docutils literal notranslate"><span class="pre">hst_7656_stis_gd71_sg140l-sg230l-g430l-g750l_o4a5_cspec.fits</span></code></p>
<p>Double check your products folder to make sure the name of your fully co-added spectra is the same as above, otherwise the subsequent cells in this notebook will not run since the pathname won’t exist. Update and run the cell below with the full co-add filename.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">stis_coadd_filename</span> <span class="o">=</span> <span class="s2">&quot;hst_7656_stis_gd71_sg140l-sg230l-g430l-g750l_o4a5_cspec.fits&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;./</span><span class="si">{</span><span class="n">stis_products_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">stis_coadd_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">hdul</span><span class="p">:</span>
    <span class="c1"># Getting the file&#39;s data</span>
    <span class="n">gd71_data</span> <span class="o">=</span> <span class="n">hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

    <span class="c1"># Getting the wavelength and flux data for the abutted file</span>
    <span class="n">wavelength</span> <span class="o">=</span> <span class="n">gd71_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">gd71_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

    <span class="c1"># Plotting the spectra</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">wavelength</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
                <span class="c1"># Setting the size of the data points</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Formatting the plot by adding titles</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;STIS GD71, abutted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

    <span class="c1"># Saving the figure to the ./stis_products_dir</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">stis_products_dir</span><span class="si">}</span><span class="s2">/gd71_stis.png&quot;</span><span class="p">)</span>

    <span class="c1"># Showing the plot below</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p><a id = cos></a></p>
</section>
<section id="putting-it-all-together-with-cos-data">
<h3>3.3 Putting it all together with COS data<a class="headerlink" href="#putting-it-all-together-with-cos-data" title="Link to this heading">#</a></h3>
<p>Let’s combine all of the STIS work from above and do another example, but this time using COS data instead of STIS. We will use the same target, GD71, and download the data from Program 11479. This has observations of GD71 using the gratings <code class="docutils literal notranslate"><span class="pre">G230L</span></code>, <code class="docutils literal notranslate"><span class="pre">G185M</span></code>, <code class="docutils literal notranslate"><span class="pre">G225M</span></code>, and <code class="docutils literal notranslate"><span class="pre">G285M</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Querying the COS data in MAST and getting the product list</span>
<span class="n">gd71_products</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
    <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
        <span class="n">proposal_id</span><span class="o">=</span><span class="mi">11479</span><span class="p">,</span>
        <span class="n">target_name</span><span class="o">=</span><span class="s2">&quot;GD71&quot;</span><span class="p">,</span>
        <span class="n">dataproduct_type</span><span class="o">=</span><span class="s2">&quot;SPECTRUM&quot;</span>
    <span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Downloading the data to the ./cos_data directory</span>
<span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
    <span class="n">gd71_products</span><span class="p">,</span>
    <span class="n">download_dir</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">cos_data_dir</span><span class="p">),</span>
    <span class="n">productSubGroupDescription</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;X1D&quot;</span><span class="p">,</span> <span class="s2">&quot;SX1&quot;</span><span class="p">]</span>
<span class="p">)</span>

<span class="c1"># Consolidating all of our files to a single directory</span>
<span class="n">consolidate_files</span><span class="p">(</span><span class="n">cos_data_dir</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we run the wrapper script:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>swrapper<span class="w"> </span>-i<span class="w"> </span>./cos_data<span class="w"> </span>-o<span class="w"> </span>./cos_products
</pre></div>
</div>
</div>
</div>
<p>Similar to our STIS example, ensure that the filename is updated to the current HASP version by running and/or updating the next cell, then plot the newly abutted COS spectra:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cos_coadd_filename</span> <span class="o">=</span> <span class="s2">&quot;hst_11479_cos_gd71_cg230l-g185m-g225m-g285m_laad_cspec.fits&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;./</span><span class="si">{</span><span class="n">cos_products_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">cos_coadd_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">hdul</span><span class="p">:</span>
    <span class="c1"># Getting the file&#39;s data</span>
    <span class="n">gd71_data</span> <span class="o">=</span> <span class="n">hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

    <span class="c1"># Getting the wavelength and flux data for the abutted file</span>
    <span class="n">wavelength</span> <span class="o">=</span> <span class="n">gd71_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">gd71_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

    <span class="c1"># Plotting the spectra</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">wavelength</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Formatting the plot</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;COS GD71, abutted&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

    <span class="c1"># Saving the figure to ./cos_products</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cos_products_dir</span><span class="si">}</span><span class="s2">/gd71_cos.png&quot;</span><span class="p">)</span>

    <span class="c1"># Showing the figure below</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We have now created and plotted the abutted COS spectra. Just for fun, let’s compare the abutted STIS and COS products!</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Getting the COS and STIS data</span>
<span class="n">stis_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">stis_products_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">stis_coadd_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">stis_data</span> <span class="o">=</span> <span class="n">stis_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

<span class="n">cos_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cos_products_dir</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">cos_coadd_filename</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">cos_data</span> <span class="o">=</span> <span class="n">cos_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

<span class="c1"># Getting the flux and wavelengths for both files</span>
<span class="n">stis_wavelength</span> <span class="o">=</span> <span class="n">stis_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
<span class="n">stis_flux</span> <span class="o">=</span> <span class="n">stis_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>

<span class="n">cos_wavelength</span> <span class="o">=</span> <span class="n">cos_data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
<span class="n">cos_flux</span> <span class="o">=</span> <span class="n">cos_data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Plotting datapoints for STIS</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">stis_wavelength</span><span class="p">,</span> <span class="n">stis_flux</span><span class="p">,</span>
            <span class="c1"># Setting the size of the datapoints</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="c1"># Setting the color for the STIS datapoints</span>
            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span>
            <span class="c1"># Adding a label for the legend</span>
            <span class="n">label</span><span class="o">=</span><span class="s2">&quot;STIS&quot;</span><span class="p">)</span>

<span class="c1"># Plotting the datapoints for COS</span>
<span class="n">plt</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">cos_wavelength</span><span class="p">,</span> <span class="n">cos_flux</span><span class="p">,</span>
            <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="s2">&quot;COS&quot;</span><span class="p">)</span>

<span class="c1"># Formatting the plot by adding labels</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;STIS &amp; COS GD71, abutted&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

<span class="c1"># Adding a legend to the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Saving the figure to our general directory</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;./gd71_cos_stis.png&quot;</span><span class="p">)</span>

<span class="c1"># Showing the plot below</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="n">stis_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="n">cos_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p><a id = download_prods></a></p>
</section>
</section>
<section id="downloading-hasp-data-products">
<h2>4. Downloading HASP Data Products<a class="headerlink" href="#downloading-hasp-data-products" title="Link to this heading">#</a></h2>
<p>As of January 2024, the HASP data products are available for download via <code class="docutils literal notranslate"><span class="pre">astroquery</span></code>. The products will also be available for query and download on the official HASP web search form (still a work in progress at the time of this notebook’s publishing).</p>
<p>Users can download coadded data themselves; we will show an example of this with a STIS coadd.</p>
<p><a id = hasp_query></a></p>
<section id="downloading-data-products-using-astroquery">
<h3>4.1 Downloading Data Products using <code class="docutils literal notranslate"><span class="pre">astroquery</span></code><a class="headerlink" href="#downloading-data-products-using-astroquery" title="Link to this heading">#</a></h3>
<p>We can use <code class="docutils literal notranslate"><span class="pre">astroquery.mast</span></code>’s <code class="docutils literal notranslate"><span class="pre">Observations</span></code> module to download data products, similar to how we downloaded our STIS and COS data in <a class="reference internal" href="#stisdownload"><span class="xref myst">Section 1.1</span></a> and <a class="reference internal" href="#cos"><span class="xref myst">Section 3.3</span></a>, respectively. Let’s download GD71 coadds for the STIS example that we created in the previous section.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">stis_query</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
    <span class="n">proposal_id</span><span class="o">=</span><span class="mi">7656</span><span class="p">,</span>
    <span class="n">target_name</span><span class="o">=</span><span class="s2">&quot;GD71&quot;</span><span class="p">,</span>
    <span class="n">dataproduct_type</span><span class="o">=</span><span class="s2">&quot;SPECTRUM&quot;</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">stis_query</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can see that our query returned a total of 11 datasets. We can see our HASP coadd is at the bottom of the table, with the value for <code class="docutils literal notranslate"><span class="pre">project</span></code> being <code class="docutils literal notranslate"><span class="pre">HASP</span></code> instead of <code class="docutils literal notranslate"><span class="pre">HST</span></code>. We will add these criteria to our query to isolate our product.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">stis_query</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
    <span class="n">proposal_id</span><span class="o">=</span><span class="mi">7656</span><span class="p">,</span>
    <span class="n">target_name</span><span class="o">=</span><span class="s2">&quot;GD71&quot;</span><span class="p">,</span>
    <span class="n">dataproduct_type</span><span class="o">=</span><span class="s2">&quot;SPECTRUM&quot;</span><span class="p">,</span>
    <span class="n">project</span><span class="o">=</span><span class="s2">&quot;HASP&quot;</span>
<span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">stis_query</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now, let’s see our product list:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">stis_prodlist</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
    <span class="n">stis_query</span>
<span class="p">)</span>

<span class="c1"># Print number of product files that can be downloaded</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Number of files: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">stis_prodlist</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Printing the entire product list</span>
<span class="n">stis_prodlist</span><span class="o">.</span><span class="n">pprint_include_names</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;productFilename&quot;</span><span class="p">,</span> <span class="s2">&quot;obs_id&quot;</span><span class="p">,</span> <span class="s2">&quot;description&quot;</span><span class="p">,</span> <span class="s2">&quot;productType&quot;</span><span class="p">,</span> <span class="s2">&quot;productSubGroupDescription&quot;</span><span class="p">,</span> <span class="s2">&quot;proposal_id&quot;</span><span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">stis_prodlist</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We have 68 files that we can download for our dataset! There are many different file types, auxiliary files such as support (<code class="docutils literal notranslate"><span class="pre">SPT</span></code>), jitter (<code class="docutils literal notranslate"><span class="pre">JIT</span></code>), wavecal exposures (<code class="docutils literal notranslate"><span class="pre">WAV</span></code>), previews of the spectra in <code class="docutils literal notranslate"><span class="pre">.PNG</span></code> format, and our science files (to name a few). Note that the files with a <code class="docutils literal notranslate"><span class="pre">productSubGroupDescription</span></code> value not equal to <code class="docutils literal notranslate"><span class="pre">CSPEC</span></code> are files that are used by <code class="docutils literal notranslate"><span class="pre">CALSTIS</span></code> during data calibration (meaning they are not produced by the HASP coadd script). The science files consist of the <code class="docutils literal notranslate"><span class="pre">SX1</span></code> files which we downloaded when running the coadd script, and the actual coadds themselves (both intermediate and final). Let’s just download the <code class="docutils literal notranslate"><span class="pre">Minimum</span> <span class="pre">Recommended</span> <span class="pre">Product</span></code> files after we create directories to store them by setting <code class="docutils literal notranslate"><span class="pre">mrp_only</span> <span class="pre">=</span> <span class="pre">True</span></code>.</p>
<p><strong>Note that the minimum recommended product filetypes are different depending on if you are downloading a HASP dataset or a non-HASP dataset, e.g. HASP will not include the <code class="docutils literal notranslate"><span class="pre">SX1</span></code> files but downloading a science dataset would.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To store our astroquery example STIS and COS data</span>
<span class="n">stis_astroquery</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./stis_astroquery/&quot;</span><span class="p">)</span>
<span class="n">stis_astroquery</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Downloading the products to our new directory</span>
<span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
    <span class="n">stis_prodlist</span><span class="p">,</span>
    <span class="n">download_dir</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">stis_astroquery</span><span class="p">),</span>
    <span class="n">mrp_only</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Putting all files to the stis_astroquery directory, rather than MAST&#39;s nested directories</span>
<span class="n">consolidate_files</span><span class="p">(</span><span class="n">stis_astroquery</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We’ve downloaded the HASP products, so let’s plot them now:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> widget

<span class="c1"># The name of the full coadd (same filename that we have in Section 3.2)</span>
<span class="n">stis_coadd_filename</span> <span class="o">=</span> <span class="s2">&quot;hst_7656_stis_gd71_sg140l-sg230l-g430l-g750l_o4a5_cspec.fits&quot;</span>
<span class="n">stis_coadd_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;./</span><span class="si">{</span><span class="n">stis_astroquery</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">stis_coadd_filename</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="n">full_coadd_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">stis_coadd_path</span><span class="p">)</span>

<span class="c1"># List of coadd files for a single grating (ignoring the full coadd above)</span>
<span class="n">grat_coadds</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;./</span><span class="si">{</span><span class="n">stis_astroquery</span><span class="si">}</span><span class="s2">/*_o4a5_*&quot;</span><span class="p">)</span> <span class="k">if</span> <span class="n">f</span> <span class="o">!=</span> <span class="n">stis_coadd_path</span><span class="p">]</span>

<span class="c1"># Starting to plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="c1"># Plotting the coadd for each grating</span>
<span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">grat_coadds</span><span class="p">:</span>
    <span class="n">file_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">file_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>
    <span class="n">wl</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">flux</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

    <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wl</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
             <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
             <span class="n">label</span><span class="o">=</span><span class="n">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    
    <span class="n">file_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    
<span class="c1"># Plotting full coadd over the plot</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">full_coadd_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

<span class="n">wl</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="n">flux</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">wl</span><span class="p">,</span> <span class="n">flux</span><span class="p">,</span>
         <span class="n">lw</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
         <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
         <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">,</span>
         <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Full Coadd&quot;</span><span class="p">)</span>

<span class="c1"># Formatting the plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;STIS GD71 Coadd from Astroquery&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

<span class="c1"># Setting ylimit to better show spectral features</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">3e-12</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="n">full_coadd_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Closing the plot to save memory</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>Click the rectangular box on the left hand side of the plot to zoom into specific portions of the plot – you will be able to see each grating’s caodd against the final full coadd. Cool!</p>
<p><a id = threshold></a></p>
</section>
</section>
<section id="changing-the-threshold-flag">
<h2>5. Changing the Threshold Flag<a class="headerlink" href="#changing-the-threshold-flag" title="Link to this heading">#</a></h2>
<p>The flux threshold flag is an optional flag that can be used to change the number of files that are co-added based on the flux of each spectra. When determining which files to abut, the script first creates a general co-add using all of the input files, and then iterates through each file to calculate the scaled median deviation. The script rejects a file if:</p>
<div class="math notranslate nohighlight">
\[
\left\langle{\frac{F_{x1d}(i) - F_{coadd(i)}}{\sigma_{x1d}(i)}}\right\rangle{} &lt; \frac{C_{thresh}}{\sqrt{N_{pix}}}
\]</div>
<p>In the equation, the median deviation from the co-add is <span class="math notranslate nohighlight">\(F_{x1d}(i) - F_{coadd(i)}\)</span>, the uncertainty per
wavelength bin of the input spectrum is <span class="math notranslate nohighlight">\(\sigma_{x1d}(i)\)</span> and <span class="math notranslate nohighlight">\(N_{pix}\)</span> is the number of wavelength bins for a given mode. Essentially, the scaled median deviation is the dispersion that quantifies the data quality (DQ) flag pixel spread for a spectrum that is less sensitive to outliers and size variations. The script will reject all files containing segments with scaled median deviation less than (i.e. more negative than) the threshold (the default is -50). For example, raising the threshold to -25 will abut less files than the default threshold of -50. Likewise, lowering the threshold to -100 will include more files in the dataset to be co-added.</p>
<p>An important thing to note about changing the threshold is that including more files increases the SNR of your co-added spectra but decreases the flux accuracy. The threshold value may be altered if a user’s science case is not dependent on the accuracy of a dataset’s absolute flux. For example, if you have a large dataset of a bright target (i.e. a standard star), then you can lower the threshold to include more files in the abutment since the spread of data quality between each file is minimal. Conversely, if you have a dataset of a dim target, or a dataset with many poor observations, then you will want to raise the threshold to only include the best data out of the dataset.</p>
<p>More information about the flux checking algorithm can be found in Section 2.3.2 of the <a class="reference external" href="https://www.stsci.edu/files/live/sites/www/files/home/hst/instrumentation/cos/documentation/instrument-science-reports-isrs/_documents/ISR2024-01.pdf">ISR COS 2024-01</a>.</p>
<p><a id = downloadthresh></a></p>
</section>
<section id="downloading-the-data">
<h2>5.1 Downloading the Data<a class="headerlink" href="#downloading-the-data" title="Link to this heading">#</a></h2>
<p>We will first download an additional dataset to illustrate how the co-add code changes abutment when the threshold flag is changed. We will download data for the spectroscopic binary, <a class="reference external" href="https://simbad.cds.unistra.fr/simbad/sim-coo?Coord=332.88126259897996+18.092873506782&amp;amp;CooFrame=FK5&amp;amp;CooEpoch=2000&amp;amp;CooEqui=2000&amp;amp;CooDefinedFrames=none&amp;amp;Radius=2&amp;amp;Radius.unit=arcmin&amp;amp;submit=submit+query&amp;amp;CoordList=">BD+17D4708</a> from Program 9631. We will run the script with three different threshold values: <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">-1,</span> <span class="pre">-2,</span> <span class="pre">-50</span></code>.</p>
<p>First we will query and download the data using <code class="docutils literal notranslate"><span class="pre">Observations</span></code> from <code class="docutils literal notranslate"><span class="pre">astroquery.mast</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Querying out data from MAST</span>
<span class="n">bd_query</span> <span class="o">=</span> <span class="n">query</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">query_criteria</span><span class="p">(</span>
            <span class="n">proposal_id</span><span class="o">=</span><span class="mi">9631</span><span class="p">,</span>
            <span class="n">target_name</span><span class="o">=</span><span class="s2">&quot;BD+17D4708&quot;</span><span class="p">,</span>
            <span class="n">dataproduct_type</span><span class="o">=</span><span class="s2">&quot;SPECTRUM&quot;</span><span class="p">,</span>
            <span class="n">provenance_name</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CALCOS&quot;</span><span class="p">,</span> <span class="s2">&quot;CALSTIS&quot;</span><span class="p">]</span>
        <span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The number of datasets queried is </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bd_query</span><span class="p">))</span><span class="si">}</span><span class="s2"> datasets. </span><span class="se">\</span>
<span class="s2">      The query list is printed below:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">bd_query</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s create a new directory and download our data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Creating the data download directories</span>
<span class="n">bd_data_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./bd_data/&quot;</span><span class="p">)</span>
<span class="n">bd_products_dir</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="s2">&quot;./bd_products/&quot;</span><span class="p">)</span>

<span class="c1"># If the directory doesn&#39;t exist, then create it</span>
<span class="n">bd_data_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">bd_products_dir</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now we get the product list and download our data to our directory <code class="docutils literal notranslate"><span class="pre">./bd_products</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Getting the product list from our query</span>
<span class="n">bd_prod</span> <span class="o">=</span> <span class="n">Observations</span><span class="o">.</span><span class="n">get_product_list</span><span class="p">(</span>
            <span class="n">bd_query</span>
        <span class="p">)</span>

<span class="c1"># Downloading the products, but only the necessary X1D and SX1 files</span>
<span class="n">Observations</span><span class="o">.</span><span class="n">download_products</span><span class="p">(</span>
            <span class="n">bd_prod</span><span class="p">,</span>
            <span class="n">download_dir</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">bd_data_dir</span><span class="p">),</span>
            <span class="n">productSubGroupDescription</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;X1D&quot;</span><span class="p">,</span> <span class="s2">&quot;SX1&quot;</span><span class="p">]</span>
        <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><a id = threshscriptrun></a></p>
</section>
<section id="running-the-co-add-script-with-different-threshold-values">
<h2>5.2 Running the Co-add Script with Different Threshold Values<a class="headerlink" href="#running-the-co-add-script-with-different-threshold-values" title="Link to this heading">#</a></h2>
<p>Let’s first organize our data so that all data files are in a single directory instead of separate observation set directories.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">consolidate_files</span><span class="p">(</span><span class="n">bd_data_dir</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now that we’ve downloaded our data and organized it to be properly run by the script, we can run the code using different threshold values. Below is a function that runs the script using inputted threshold values, which will allow us to run the code in a single cell rather than multiple.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">run_script</span><span class="p">(</span><span class="n">indir</span><span class="p">,</span> <span class="n">outdir</span><span class="p">,</span> <span class="n">thresh</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Run the hasp wrapper script with different thresholds.</span>
<span class="sd">    -------------------</span>
<span class="sd">    Input:</span>
<span class="sd">    str indir : Directory that holds all un-abutted data.</span>
<span class="sd">    str outdir : Directory that will store output folders and files.</span>
<span class="sd">    list thresh : List of threshold values used for script.</span>
<span class="sd">    --------------------</span>
<span class="sd">    Output:</span>
<span class="sd">    Wrapper script is run using different threshold values.</span>
<span class="sd">    A folder will be created for each value used, containing abutted products.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># Looping through thresh vals to run through script</span>
    <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">thresh</span><span class="p">:</span>
        <span class="c1"># Creating a folder for each thresh val</span>
        <span class="n">output</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;out&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">val</span><span class="p">))</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">output</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">output</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Running script where threshold = </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">val</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Specifying the arguments for the scipt</span>
        <span class="n">arguments</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-i&quot;</span><span class="p">,</span> <span class="n">indir</span><span class="p">,</span> <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">output</span><span class="p">,</span> <span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="n">val</span><span class="p">]</span>

        <span class="c1"># Creating a list composed of the script commands and arguments</span>
        <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;python&quot;</span><span class="p">,</span> <span class="s2">&quot;-m&quot;</span><span class="p">,</span> <span class="s2">&quot;hasp.wrapper&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">arguments</span>

        <span class="c1"># Running the script with the arguments above</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">command</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s create a variable called <code class="docutils literal notranslate"><span class="pre">thresholds</span></code> that will consist of the different threshold values that we will use to help us visualize the relationship between number of files co-added and the final co-added product. We will run the script using each of these values.</p>
<p><em>Note: feel free to run the cells with more threshold values!</em></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remember that the default value is t = -50</span>
<span class="n">thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-1&quot;</span><span class="p">,</span> <span class="s2">&quot;-2&quot;</span><span class="p">,</span> <span class="s2">&quot;-50&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>Let’s run the script on our data now using the <code class="docutils literal notranslate"><span class="pre">run_script</span></code> function, using the thresholds defined above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">run_script</span><span class="p">(</span><span class="n">bd_data_dir</span><span class="p">,</span> <span class="n">bd_products_dir</span><span class="p">,</span> <span class="n">thresholds</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><a id = analyzethresh></a></p>
</section>
<section id="analyzing-the-different-co-added-spectra-of-different-threshold-values">
<h2>5.3 Analyzing the Different Co-added Spectra of Different Threshold Values<a class="headerlink" href="#analyzing-the-different-co-added-spectra-of-different-threshold-values" title="Link to this heading">#</a></h2>
<p>Now that we’ve run the script on our dataset using differnet threshold values, we can start to analyze the differences between them.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">run_script</span></code> function created multiple folders for the abutted products; one folder for each threshold value (ex. <code class="docutils literal notranslate"><span class="pre">./out-50</span></code>). In each folder is the fully co-added spectra. At the time of this notebook, the fully co-added filename is:</p>
<p><code class="docutils literal notranslate"><span class="pre">hst_9631_stis_bdp17d4708_g230lb-g430l-g750l_o8h1_cspec.fits</span></code></p>
<p>Before you continue, make sure that the fully co-added filename is valid, as the filenaming structure may have changed since this notebook was published. Update and run the cell below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bd_coadd_name</span> <span class="o">=</span> <span class="s2">&quot;hst_9631_stis_bdp17d4708_g230lb-g430l-g750l_o8h1_cspec.fits&quot;</span>
</pre></div>
</div>
</div>
</div>
<p>We can see how many files were used in the co-add by running next the cell.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Iterating through the threshold values</span>
<span class="k">for</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="n">thresholds</span><span class="p">:</span>
    <span class="c1"># Path to the co-added file</span>
    <span class="n">path_to_coadd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/out</span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">bd_coadd_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">coadd_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">path_to_coadd</span><span class="p">)</span>

    <span class="c1"># Opening the file to check num files, and printing</span>
    <span class="n">numfiles</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">coadd_hdul</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;FILENAME&quot;</span><span class="p">])</span>

    <span class="c1"># Printing the number of files for the given threshold value</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The number of files co-added for t = </span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2"> is </span><span class="si">{</span><span class="n">numfiles</span><span class="si">}</span><span class="s2"> files&quot;</span><span class="p">)</span>

    <span class="n">coadd_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>As we can see, as we increase the threshold value (<code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">-50</span></code> -&gt; <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">-1</span></code>) we went from 12 files being used to only 6. Since we are using more data with the lower threshold (<code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">-50</span></code>), we should see an increase in SNR.</p>
<section id="looking-at-snr">
<h3>Looking at SNR:<a class="headerlink" href="#looking-at-snr" title="Link to this heading">#</a></h3>
<p>We’ll plot the SNR of both coadds below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="c1"># A list of the the non-default thresholds used.</span>
<span class="n">nondefault_thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="n">thresh</span> <span class="k">for</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="n">thresholds</span> <span class="k">if</span> <span class="n">thresh</span> <span class="o">!=</span> <span class="s2">&quot;-50&quot;</span><span class="p">]</span>

<span class="c1"># List of colors to be used when plotting</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;blue&#39;</span><span class="p">,</span>
          <span class="s1">&#39;red&#39;</span><span class="p">]</span>

<span class="c1"># Creating our figure: 1 column and len(thresholds) rows</span>
<span class="n">figure</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nondefault_thresholds</span><span class="p">),</span> <span class="mi">1</span><span class="p">,</span>
                          <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>

<span class="c1"># Iterating through threshold values to compare their SNR to default output</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">nondefault_thresholds</span><span class="p">):</span>

    <span class="c1"># The SNR and wavelength data for the default threshold co-add output</span>
    <span class="n">default_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/out-50/</span><span class="si">{</span><span class="n">bd_coadd_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">default_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">default_path</span><span class="p">)</span>
    <span class="n">data50</span> <span class="o">=</span> <span class="n">default_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>
    <span class="n">wavelength50</span> <span class="o">=</span> <span class="n">data50</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">snr50</span> <span class="o">=</span> <span class="n">data50</span><span class="p">[</span><span class="s2">&quot;SNR&quot;</span><span class="p">]</span>

    <span class="c1"># Getting wavelength and SNR for the current iteration&#39;s threshold value</span>
    <span class="n">curr_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/out</span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">bd_coadd_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">curr_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">curr_path</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">curr_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>
    <span class="n">wavelength</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">snr</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s2">&quot;SNR&quot;</span><span class="p">]</span>

    <span class="c1"># This is the number of files used in the particular co-add</span>
    <span class="n">numfiles</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">curr_path</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;FILENAME&quot;</span><span class="p">])</span>
    <span class="n">numfiles_default</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">default_path</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;FILENAME&quot;</span><span class="p">])</span>

    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">wavelength50</span><span class="p">,</span> <span class="n">snr50</span><span class="p">,</span>
                  <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;t = -50, </span><span class="si">{</span><span class="n">numfiles_default</span><span class="si">}</span><span class="s2"> files co-added&quot;</span><span class="p">,</span>
                  <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                  <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">)</span>

    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">wavelength</span><span class="p">,</span> <span class="n">snr</span><span class="p">,</span>
                  <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;t = </span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">numfiles</span><span class="si">}</span><span class="s2"> files co-added&quot;</span><span class="p">,</span>
                  <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                  <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                  <span class="n">alpha</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>

    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="c1"># Highlighting the region of the plot with the biggest SNR change</span>
    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">axvspan</span><span class="p">(</span><span class="mi">3050</span><span class="p">,</span> <span class="mi">5700</span><span class="p">,</span>
                  <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span>
                  <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">)</span>

    <span class="c1"># Adding formatting</span>
    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;SNR vs Wavelength for t = </span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">, -50&quot;</span><span class="p">,</span>
                    <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
    <span class="n">ax</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;SNR&quot;</span><span class="p">)</span>

    <span class="n">default_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">curr_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="n">figure</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

<span class="c1"># Saving the figure to ./bd_products</span>
<span class="n">figure</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/snr_wavelength.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can clearly see above that using more files in the co-add substantially increases SNR. This is especially prevalent in the grey highlighted section of the top plot, where we set <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">-1</span></code>. The number of files used with that threshold value were 6 files, whereas the default threshold value of <code class="docutils literal notranslate"><span class="pre">t</span> <span class="pre">=</span> <span class="pre">-50</span></code> used 12 files.</p>
</section>
<section id="looking-at-flux">
<h3>Looking at Flux:<a class="headerlink" href="#looking-at-flux" title="Link to this heading">#</a></h3>
<p>We’re going to create two diffrent plots to analyze our flux data: a standard flux vs wavelength plot, and a differential plot. We will also calculate the percent change in flux between the threshold values vs the default.</p>
<p>A differential plot shows the absolute differences between two datasets; in our case, we will be seeing the flux difference at each wavelength between one threshold value’s co-add and the default co-add. However, we will first need to interpolate the spectra to a common wavelength grid. Essentially, we will be adjusting the wavelength values so the two datasets can share the same wavelength points.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Thresholds used that aren&#39;t the default</span>
<span class="n">nondefault_thresholds</span> <span class="o">=</span> <span class="p">[</span><span class="n">thresh</span> <span class="k">for</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="n">thresholds</span> <span class="k">if</span> <span class="n">thresh</span> <span class="o">!=</span> <span class="s2">&quot;-50&quot;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">thresh</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">nondefault_thresholds</span><span class="p">):</span>
    <span class="c1"># Getting the data for the co-add that uses default threshold of t = -50</span>
    <span class="n">default_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/out-50/</span><span class="si">{</span><span class="n">bd_coadd_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">default_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">default_path</span><span class="p">)</span>
    <span class="n">thresh_data50</span> <span class="o">=</span> <span class="n">default_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

    <span class="n">wavelength50</span> <span class="o">=</span> <span class="n">thresh_data50</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">flux50</span> <span class="o">=</span> <span class="n">thresh_data50</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>

    <span class="c1"># Getting data for the co-add that uses the current iteration&#39;s threshold</span>
    <span class="n">curr_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/out</span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">bd_coadd_name</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">curr_hdul</span> <span class="o">=</span> <span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">curr_path</span><span class="p">)</span>
    <span class="n">thresh_data_curr</span> <span class="o">=</span> <span class="n">curr_hdul</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">data</span>

    <span class="n">wavelength_curr</span> <span class="o">=</span> <span class="n">thresh_data_curr</span><span class="p">[</span><span class="s2">&quot;WAVELENGTH&quot;</span><span class="p">]</span>
    <span class="n">flux_curr</span> <span class="o">=</span> <span class="n">thresh_data_curr</span><span class="p">[</span><span class="s2">&quot;FLUX&quot;</span><span class="p">]</span>

    <span class="c1"># Getting the number of files used in co-add, will put this in the plot</span>
    <span class="n">numfiles</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">curr_path</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;FILENAME&quot;</span><span class="p">])</span>
    <span class="n">numfiles_default</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">fits</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">default_path</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">data</span><span class="p">[</span><span class="s2">&quot;FILENAME&quot;</span><span class="p">])</span>

    <span class="c1"># Minimum wavelength value in dataset</span>
    <span class="n">minwave</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">wavelength50</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span> <span class="n">wavelength_curr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">())</span>

    <span class="c1"># Maximum wavelength value in dataset</span>
    <span class="n">maxwave</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">wavelength50</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(),</span> <span class="n">wavelength_curr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>

    <span class="c1"># Creating a common wavelength grid using shape of default wavelength axis</span>
    <span class="n">common_wavelength</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">start</span><span class="o">=</span><span class="n">minwave</span><span class="p">,</span>
                                  <span class="n">stop</span><span class="o">=</span><span class="n">maxwave</span><span class="p">,</span>
                                  <span class="n">step</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Interpolating the default + current threshold co-add onto new grid</span>
    <span class="n">interp_flux50</span> <span class="o">=</span> <span class="n">interp1d</span><span class="p">(</span><span class="n">wavelength50</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                             <span class="n">flux50</span><span class="p">,</span>
                             <span class="n">kind</span><span class="o">=</span><span class="s1">&#39;linear&#39;</span><span class="p">,</span>
                             <span class="n">fill_value</span><span class="o">=</span><span class="s2">&quot;extrapolate&quot;</span><span class="p">)(</span><span class="n">common_wavelength</span><span class="p">)</span>
    
    <span class="n">interp_flux_curr</span> <span class="o">=</span> <span class="n">interp1d</span><span class="p">(</span><span class="n">wavelength_curr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                <span class="n">flux_curr</span><span class="p">,</span>
                                <span class="n">kind</span><span class="o">=</span><span class="s1">&#39;linear&#39;</span><span class="p">,</span>
                                <span class="n">fill_value</span><span class="o">=</span><span class="s2">&quot;extrapolate&quot;</span><span class="p">)(</span><span class="n">common_wavelength</span><span class="p">)</span>

    <span class="c1"># Creating two subplots, one flux vs wavelength and one differential plot</span>
    <span class="n">fig</span><span class="p">,</span> <span class="p">[</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">]</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span>
                                   <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">))</span>

    <span class="c1"># Plotting the top plot, a.k.a. the flux vs wavelength for t = -50</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">common_wavelength</span><span class="p">,</span> <span class="n">interp_flux50</span><span class="p">,</span>
                <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;t = -50, </span><span class="si">{</span><span class="n">numfiles_default</span><span class="si">}</span><span class="s2"> files&quot;</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Plotting flux vs wavelength on same plot for current co-add</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">common_wavelength</span><span class="p">,</span> <span class="n">interp_flux_curr</span><span class="p">,</span>
                <span class="n">label</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;t = </span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">, </span><span class="si">{</span><span class="n">numfiles</span><span class="si">}</span><span class="s2"> files&quot;</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>

    <span class="c1"># Calculating difference between current co-add and default flux values</span>
    <span class="n">flux_diff</span> <span class="o">=</span> <span class="n">interp_flux_curr</span> <span class="o">-</span> <span class="n">interp_flux50</span>

    <span class="c1"># Plotting the differential plot</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">common_wavelength</span><span class="p">,</span> <span class="n">flux_diff</span><span class="p">,</span>
                <span class="n">color</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span>
                <span class="n">s</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Calculating the percent difference and putting it on the plot</span>
    <span class="n">percent_difference</span> <span class="o">=</span> <span class="nb">abs</span><span class="p">(</span><span class="n">interp_flux50</span> <span class="o">-</span> <span class="n">interp_flux_curr</span><span class="p">)</span> <span class="o">/</span> <span class="p">((</span><span class="n">interp_flux50</span> <span class="o">+</span> <span class="n">interp_flux_curr</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>

    <span class="c1"># Calculating the mean percent difference (ignoring NaNs)</span>
    <span class="n">percent_difference</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">percent_difference</span><span class="p">),</span> <span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Adding text box onto plot that displays newly calculated 5 difference</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.02</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">,</span>
             <span class="sa">f</span><span class="s2">&quot;Average percent difference: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">percent_difference</span><span class="p">)</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">,</span>
             <span class="n">transform</span><span class="o">=</span><span class="n">ax0</span><span class="o">.</span><span class="n">transAxes</span><span class="p">,</span>
             <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">,</span>
             <span class="n">bbox</span><span class="o">=</span><span class="nb">dict</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
                       <span class="n">alpha</span><span class="o">=</span><span class="mf">0.3</span><span class="p">))</span>
    
    <span class="c1"># Setting the y axis limits on both plots to be the same</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1e-12</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1e-12</span><span class="p">)</span>

    <span class="c1"># Adding formatting</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Flux vs Wavelength for t = </span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">, -50&quot;</span><span class="p">,</span>
                  <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>

    <span class="n">ax0</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

    <span class="c1"># Putting the legend on the flux plot</span>
    <span class="n">ax0</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

    <span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Absolute Difference for t = </span><span class="si">{</span><span class="n">thresh</span><span class="si">}</span><span class="s2">, -50&quot;</span><span class="p">,</span>
                  <span class="n">weight</span><span class="o">=</span><span class="s2">&quot;bold&quot;</span><span class="p">)</span>
    
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Wavelength [$\AA$]&#39;</span><span class="p">)</span>
    <span class="n">ax1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;Flux Difference [$erg\ s^{-1}\ cm^{-2}\ \AA^{-1}$]&#39;</span><span class="p">)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bd_products_dir</span><span class="si">}</span><span class="s2">/flux_differential.png&quot;</span><span class="p">)</span>

    <span class="n">fig</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

    <span class="n">default_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">curr_hdul</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="congrats-on-completing-the-notebook">
<h1>Congrats on completing the notebook!<a class="headerlink" href="#congrats-on-completing-the-notebook" title="Link to this heading">#</a></h1>
<section id="there-are-more-tutorial-notebooks-for-custom-co-addition-cases-in-this-repo-check-them-out">
<h2>There are more tutorial notebooks for custom co-addition cases in <a class="reference external" href="https://github.com/spacetelescope/hst_notebooks/tree/main/notebooks/HASP">this repo</a>, check them out!<a class="headerlink" href="#there-are-more-tutorial-notebooks-for-custom-co-addition-cases-in-this-repo-check-them-out" title="Link to this heading">#</a></h2>
</section>
<section id="about-this-notebook">
<h2>About this Notebook<a class="headerlink" href="#about-this-notebook" title="Link to this heading">#</a></h2>
<p><strong>Author:</strong> Sierra Gomez (sigomez&#64;stsci.edu)</p>
<p><strong>Updated on:</strong> 01/29/2024</p>
<p><em>This tutorial was generated to be in compliance with the <a class="reference external" href="https://github.com/spacetelescope/style-guides">STScI style guides</a> and would like to cite the <a class="reference external" href="https://github.com/spacetelescope/style-guides/blob/master/templates/example_notebook.ipynb">Jupyter guide</a> in particular.</em></p>
</section>
<section id="citations">
<h2>Citations<a class="headerlink" href="#citations" title="Link to this heading">#</a></h2>
<p>If you use the following packages for published research, please cite the authors. Follow these links for more information about citations:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://docs.astropy.org/en/stable/index.html">Citing <code class="docutils literal notranslate"><span class="pre">astropy</span></code></a></p></li>
<li><p><a class="reference external" href="https://astroquery.readthedocs.io/en/latest/">Citing <code class="docutils literal notranslate"><span class="pre">astroquery</span></code></a></p></li>
<li><p><a class="reference external" href="https://matplotlib.org/stable/users/project/citing.html">Citing <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code></a></p></li>
<li><p><a class="reference external" href="https://numpy.org/citing-numpy/">Citing <code class="docutils literal notranslate"><span class="pre">numpy</span></code></a></p></li>
<li><p><a class="reference external" href="https://scipy.org/citing-scipy/">Citing <code class="docutils literal notranslate"><span class="pre">scipy</span></code></a></p></li>
</ul>
<a class="reference internal image-reference" href="https://raw.githubusercontent.com/spacetelescope/notebooks/master/assets/stsci_pri_combo_mark_horizonal_white_bkgd.png"><img alt="Space Telescope Logo" src="https://raw.githubusercontent.com/spacetelescope/notebooks/master/assets/stsci_pri_combo_mark_horizonal_white_bkgd.png" style="width: 200px;" />
</a>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "spacetelescope/hst_notebooks",
            ref: "main",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./notebooks/HASP/CoaddTutorial"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../Setup/Setup.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Installing the Hubble Advanced Spectral Products Script</p>
      </div>
    </a>
    <a class="right-next"
       href="../FluxScaleTutorial/FluxScaleTutorial.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Scaling Flux while using the Hubble Advanced Spectral Products Script</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">Inputting User Data using the Hubble Advanced Spectral Products Script</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#this-notebook-is-designed-to-walk-you-through-downloading-and-using-the-hubble-advanced-spectral-products-hasp-co-add-script"><span style="font-weight:normal">This Notebook is designed to walk you through downloading and using the <strong>Hubble Advanced Spectral Products (HASP)</strong> co-add script.</span></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-goals">Learning Goals:</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#table-of-contents">Table of Contents</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#introduction">0. Introduction</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#imports">Imports</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-hst-spectroscopic-data">1. Downloading HST Spectroscopic Data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#using-astroquery-to-download-stis-data">1.1 Using <code class="docutils literal notranslate"><span class="pre">Astroquery</span></code> to Download STIS Data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-the-co-add-script">2. Running the Co-add Script</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#working-with-co-added-data-products">3. Working with Co-added Data Products</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#understanding-the-output-files">3.1 Understanding the Output Files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#viewing-the-co-added-data">3.2 Viewing the Co-added Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#putting-it-all-together-with-cos-data">3.3 Putting it all together with COS data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-hasp-data-products">4. Downloading HASP Data Products</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-data-products-using-astroquery">4.1 Downloading Data Products using <code class="docutils literal notranslate"><span class="pre">astroquery</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#changing-the-threshold-flag">5. Changing the Threshold Flag</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#downloading-the-data">5.1 Downloading the Data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#running-the-co-add-script-with-different-threshold-values">5.2 Running the Co-add Script with Different Threshold Values</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analyzing-the-different-co-added-spectra-of-different-threshold-values">5.3 Analyzing the Different Co-added Spectra of Different Threshold Values</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#looking-at-snr">Looking at SNR:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#looking-at-flux">Looking at Flux:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#congrats-on-completing-the-notebook">Congrats on completing the notebook!</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#there-are-more-tutorial-notebooks-for-custom-co-addition-cases-in-this-repo-check-them-out">There are more tutorial notebooks for custom co-addition cases in this repo, check them out!</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#about-this-notebook">About this Notebook</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#citations">Citations</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By STScI
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022-2024.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>